spec/upstream_spec.lua

```diff
@@ -0,0 +1,152 @@
+    local valid_url = 'http://localhost:8080/path?query'
+
+    describe('.new', function()
+        it('returns an error on invalid upstream', function()
+            assert.returns_error('invalid upstream', Upstream.new('invalid uri'))
+        end)
+
+        it('returns new instance', function()
+            assert(Upstream.new(valid_url))
+        end)
+
+        it('returns table with location_name', function()
+            -- @upstream location is defined in the apicast.conf file
+            assert.equal('@upstream', Upstream.new(valid_url).location_name)
+        end)
+
+        it('returns table with upstream_name', function()
+            -- upstream name is defined in the upstream.conf file
+            assert.equal('upstream', Upstream.new(valid_url).upstream_name)
+        end)
+
+        it('returns table with resolver', function()
+            assert.equal(require('resty.resolver'), Upstream.new(valid_url).resolver)
+        end)
+    end)
+
+    describe(':resolve', function()
+        it('uses .resolver', function()
+            local upstream = Upstream.new(valid_url)
+            local resolver, instance, servers = {}, {}, {}
+            stub(resolver, 'instance').returns(instance)
+            stub(instance, 'get_servers').returns(servers)
+
+            upstream.resolver = resolver
+
+            assert.equal(servers, upstream:resolve())
+            assert.spy(instance.get_servers).was_called_with(instance, 'localhost', match.contains({ port = 8080 }))
+        end)
+
+        -- this is useful when we want to pre populate resolved servers in tests
+        it('keeps and returns previously resolved servers', function()
+            local upstream = Upstream.new(valid_url)
+            local servers = { 'resolved servers instance' }
+            upstream.servers = servers
+
+            assert.equal(upstream.servers, upstream:resolve())
+            assert.equal(upstream.servers, servers)
+        end)
+    end)
+
+    describe(':port', function()
+        it('returns port from the URI', function()
+            assert.same(8090, Upstream.new('http://host:8090'):port())
+        end)
+
+        it('returns default port for the scheme when none is provided', function()
+            assert.same(443, Upstream.new('https://example.com'):port())
+        end)
+    end)
+
+    describe(':rewrite', function()
+        it('sets request Host header to the URI host', function()
+            stub(ngx.req, 'set_header')
+
+            assert(Upstream.new('http://example.com/')):rewrite()
+
+            assert.spy(ngx.req.set_header).was_called_with('Host', 'example.com')
+        end)
+
+        it('sets request Host header to the host passed as argument', function()
+            stub(ngx.req, 'set_header')
+
+            assert(Upstream.new('http://example.com/')):rewrite('localhost')
+
+            assert.spy(ngx.req.set_header).was_called_with('Host', 'localhost')
+        end)
+    end)
+
+    describe(':rewrite', function()
```

Comment:
We should test the case where `ngx.headers_sent` is true.

Reply:
Yep.